src/main/java/filter/expression/Qualifier.java

```diff
@@ -136,28 +134,96 @@ public static boolean process(IModel model, FilterExpression expr, TurboIssue is
+            TurboMilestone openMilestone = openMilestones.get(0);
+            boolean hasDueDate = openMilestone.getDueDate().isPresent();
+            if (!hasDueDate) {
+                milestones.add(openMilestone);
+            }
         }
 
-        int currentIndex = 0;
+        return milestones;
+    }
 
-        for (TurboMilestone checker : allMilestones) {
-            boolean overdue = checker.getDueDate().isPresent() &&
-                    checker.getDueDate().get().isBefore(LocalDate.now());
-            boolean relevant = !(overdue && checker.getOpenIssues() == 0);
+    /**
+     * Returns a map where the key denotes offset from "current" milestone.
+     * i.e. 0 for "current" milestone, -1 for "current-1" milestone, etc.
+     */
+    private static Map<Integer, TurboMilestone> getMilestoneAliasIndex(List<TurboMilestone> milestones) {
+        List<TurboMilestone> sortedMilestones = TurboMilestone.sortByDueDate(milestones);
+        Optional<Integer> currentMilestoneIndex = getCurrentMilestoneIndex(sortedMilestones);
 
-            if (checker.isOpen() && relevant) {
-                return Optional.of(currentIndex);
-            }
+        assert currentMilestoneIndex.isPresent() || !currentMilestoneIndex.isPresent() && sortedMilestones.isEmpty();
 
-            currentIndex++;
+        return IntStream
+                .range(0, sortedMilestones.size())
+                .boxed()
+                .collect(Collectors.toMap(
+                        i -> i - currentMilestoneIndex.get(),
+                        i -> sortedMilestones.get(i)));
+    }
+
+    /**
+     * "Current" milestone is an ongoing milestone with the earliest due date. However, if there
+     * is only one open milestone, it will be considered as the "current" milestone, even if it
+     * does not have due date.
+     *
+     * This method expects the milestone list to be sorted (by due date)
+     */
+    private static Optional<Integer> getCurrentMilestoneIndex(List<TurboMilestone> sortedMilestones) {
+        if (sortedMilestones.isEmpty()) {
+            return Optional.empty();
+        }
+
+        int openMilestonesCount = TurboMilestone.getOpenMilestones(sortedMilestones).size();
+        if (openMilestonesCount == 1) {
+            return getFirstOpenMilestonePosition(sortedMilestones);
         }
 
-        // if no open milestone, set current as one after last milestone
-        // - this means that no such milestone, which will return no issue
-        return Optional.of(allMilestones.size());
+        // Look for the first milestone in the (sorted) list that is ongoing.
+        Optional<Integer> firstOngoingMilestonePosition = getFirstOngoingMilestonePosition(sortedMilestones);
+
+        // if no ongoing milestone, set current as one after last milestone
+        // - this means that there is no "current" milestone, but it is possible
+        // that there is a milestone before the "current" one (i.e. current-[n]
+        // where n >= 1)
```

Comment:
Anyway, I'm a bit uneasy about this behavior. It seems beyond the advertised purpose of this method, which is to find the current milestone. How does the caller know the returned value has this special meaning? Perhaps we should simply return `Optional.empty()` and let the caller figure out what to do?